File: src/xmlvm2js/org/xmlvm/client/Timer.js
/* Copyright (c) 2002-2011 by XMLVM.org
 *
 * Project Info:  http://www.xmlvm.org
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
 * USA.
 */

qx.Class.define("org_xmlvm_client_Timer", {
  extend: java_lang_Object,
  construct: function() {
  	this.listeners = new Array();
  },
  members:
  {
  	listeners: 0,
  	$addActionListener___java_awt_event_ActionListener: function(listener) {
  		this.listeners.push(listener);
  	},
  	$runOnce___int: function(millis) {
  		var timer = this;
		setTimeout(function() {
			timer.fire();
		}, millis);  	
  	},
  	$runRepeating___int: function(millis) {
  		var timer = this;
		setInterval(function() {
			timer.fire();
		}, millis);  	
  	},
  	fire: function() {
  		for(i=0; i<this.listeners.length; ++i) {
  			this.listeners[i].$actionPerformed___java_awt_event_ActionEvent(new java_lang_null());
  		}
  	}
  }
});

File: src/xmlvm/org/xmlvm/client/Timer.java
/*
 * Copyright (c) 2004-2010 XMLVM --- An XML-based Programming Language
 * 
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 * 
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 * 
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 675 Mass
 * Ave, Cambridge, MA 02139, USA.
 * 
 * For more information, visit the XMLVM Home Page at http://www.xmlvm.org
 */

package org.xmlvm.client;

import java.awt.event.ActionListener;
import java.util.ArrayList;
import java.util.List;

public class Timer extends Thread {

    private List<ActionListener> listeners = new ArrayList<ActionListener>();
    private int                  millis;
    private boolean              repeating = true;


    @Override
    public void run() {
        while (repeating && !isInterrupted()) {
            try {
                sleep(millis);
            } catch (InterruptedException e) {
                this.interrupt();
            }
            for (ActionListener listener : listeners) {
                listener.actionPerformed(null);
            }
        }
    }

    public void addActionListener(ActionListener listener) {
        this.listeners.add(listener);
    }

    public void runOnce(int millis) {
        this.millis = millis;
        repeating = false;
        start();
    }

    public void runRepeating(int millis) {
        this.millis = millis;
        repeating = true;
        start();
    }
}
